rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2244 @@
+
+`slog` is a logging _framework_. It offers all the fundamental tools needed out-of-the-box to capture log records, define and implement the composable pieces of a logging pipeline, and pass them through that pipeline to an eventual destination. It has conventions and trade-offs baked into the design of its API. Loggers are treated explicitly as values in data structures and as arguments, and callers can control whether to pass owned or borrowed data.
+
+`log` is a logging _facade_. It's only concerned with a standard, minimal API for capturing log records, and surfacing those records to some consumer. The tools provided by `log` are only those that are fundamental to the operation of the `log!` macro. From `log`'s point of view, a logging framework like `slog` is a black-box implementation of the `Log` trait. In this role, the `Log` trait can act as a common entrypoint for capturing log records. That means the `Record` type can act as a common container for describing a log record. `log` has its own set of trade-offs baked into the design of its API. The `log!` macro assumes a single, global entrypoint, and all data in a log record is borrowed from the callsite.
+
+A healthy logging ecosystem needs both `log` and frameworks like `slog`. As a standard API, `log` can support a diverse but cohesive ecosystem of logging tools in Rust by acting as the glue between libraries, frameworks, and applications. A lot of libraries already depend on it. In order to really fulfil this role though, `log` needs to support structured logging so that libraries and their consumers can take advantage of it in a framework-agnostic way.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Logging structured key-value pairs
+
+Structured logging is supported in `log` by allowing typed key-value pairs to be associated with a log record. A `;` separates structured key-value pairs from values that are replaced into the message:
+
+```rust
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+Any `value` or `key = value` expressions before the `;` in the macro will be interpolated into the message as unstructured text using `std::fmt`. This is the `log!` macro we have today. Any `value` or `key = value` expressions after the `;` will be captured as structured key-value pairs. These structured key-value pairs can be inspected or serialized, retaining some notion of their original type. That means in the above example, the `message` key is unstructured, and the `correlation` and `user` keys are structured:
+
+```
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    ^^^^^^^^^^^^^^^^^^^
+    unstructured
```

Comment:
This surprised me.  I'm use to <small>(not in rust)</small> doing messages like `"action {action_name} completed in {duration}ms"`, and still getting the properties as real columns that I can filter and pivot on.

Reply:
Yes, that's a better way to do things, but we couldn't do this and maintain potential structure without potential breakage to callers of `log!`, because the trait bounds within that format would change.